provider/parameter.go

```diff
@@ -144,38 +138,23 @@ func parameterDataSource() *schema.Resource {
 				return diag.Errorf("ephemeral parameter requires the default property")
 			}
 
-			// TODO: Should we move this into the Valid() function on
-			//  Parameter?
-			if len(parameter.Validation) == 1 {
-				validation := &parameter.Validation[0]
-				err = validation.Valid(parameter.Type, value)
-				if err != nil {
-					return diag.FromErr(err)
-				}
-			}
-
-			// Validate options
-			_, parameter.FormType, err = ValidateFormType(parameter.Type, len(parameter.Option), parameter.FormType)
-			if err != nil {
-				return diag.Diagnostics{
-					{
-						Severity:      diag.Error,
-						Summary:       "Invalid form_type for parameter",
-						Detail:        err.Error(),
-						AttributePath: cty.Path{cty.GetAttrStep{Name: "form_type"}},
-					},
-				}
+			var input *string
+			envValue, ok := os.LookupEnv(ParameterEnvironmentVariable(parameter.Name))
+			if ok {
+				input = &envValue
 			}
-			// Set the form_type back in case the value was changed.
-			// Eg via a default. If a user does not specify, a default value
-			// is used and saved.
-			rd.Set("form_type", parameter.FormType)
 
-			diags := parameter.Valid()
+			value, diags := parameter.ValidateInput(input)
 			if diags.HasError() {
 				return diags
 			}
 
+			// Always set back the value, as it can be sourced from the default
+			rd.Set("value", value)
+
+			// Set the form_type as it could have changed in the validation.
```

Comment:
I found this curious and confusing. A fuller explanation is probably worth setting out here: the `form_type` can start off unset, in which case the "validation" code selects a form type based on the `type`.

It's also a bit of a naming failure that we call the procedure that does this "Validation", which implies it doesn't change anything and just renders a verdict. In the Coderd code, I believe we called this procedure "resolution".

Reply:
It is unfortunate. I updated the comment.